### DISCUSSION TITLE
Script to remove all interfaces, change tenant, change location

### DISCUSSION THREAD

**Discussion by domainzero** [netbox-community/netbox#15144] [Help Wanted!] [answered]
Hey all,

I'm working on a script to prepare hosts for decommissioning. Part of our process for doing that is removing all cables, changing the tenant, and moving the host to a storage location.

To date, I have been able to do all of this via Python script that uses the API via HTTP requests, but I want to move this into the Scripts pane in NetBox and have a GUI to accompany it so members of my team can have easier access to doing this.

Are there any functions or anything I should be aware of that can aide me in simply deleting all cables connected to interfaces on a given host? I'm sure I can figure out setting the tenant and location otherwise, but recursively removing cables is where I've been getting stuck.

**Comment by ZephyrusBlaze** [comment]


**Reply by candlerb** [reply]
I think it's very likely that the above answer was garbage spewed out of ChatGPT, despite not being labelled as such.

Firstly, I strongly recommend you DON'T take this approach. Calling out from within a custom script to the REST API loses all of the transactional guarantees that Scripts normally have; it introduces more opportunity for errors and timeouts; and it means you have to hard-code API credentials within your script. In any case, the OP already said that they knew how to do it using the REST API, so it completely missed the point of the question.

But secondly, the construct `Q(termination_a_id=interface.id) | Q(termination_b_id=interface.id))` is from the Django ORM, not from the REST API.  This code blindly mixes two completely different approaches, and it could not possibly work as written - it does not even import `Q`.  And even if it did work, it would be for the cable model prior to Netbox [v3.3.0](https://github.com/netbox-community/netbox/releases/v3.3.0).

Please don't paste ChatGPT output without labelling it as such - it just wastes everyone's time.

**Reply by ZephyrusBlaze** [reply]
Oh, yea I really dunno anything about netbox, but yk I wanted to earn the galaxy brain achievement but unfortunately GitHub has made it unachievable in the community discussions so I had to answer on random repos. I'll make sure to label my future answers as AI generated. I apologize for the inconvenience.

**Comment by candlerb** [comment, marked as answer]
> Are there any functions or anything I should be aware of that can aide me in simply deleting all cables connected to interfaces on a given host? I'm sure I can figure out setting the tenant and location otherwise, but recursively removing cables is where I've been getting stuck.

You can iterate over the interfaces on a device, and then delete the cable attached to that interface (there can only ever be one per interface). I don't think there's any need for recursion, unless you want to follow a chain of cables through patch panels etc and delete *all* of those cables - is that the case?

A good way to try out Django ORM code interactively is within [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/), which is most easily started using

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
```

Then you can experiment:

```
>>> d = Device.objects.get(name="gw2")
>>> intf = d.interfaces.get(name="ether8")
>>> help(Interface)
...
>>> intf.cable
<Cable: #51>
```

If that's all you need, then you can simply delete that cable. Iteration is just a question of looping over the interfaces:

```
>>> for intf in d.interfaces.all(): print(intf)
...
```

If you're using cables with multiple terminations though, you might want to delete just the termination into that interface, rather than the entire cable (which deletes all terminations).  However, doing so risks leaving "orphaned" cables which are not connected at one or both ends.  There is a WIP script [here](https://github.com/netbox-community/customizations/pull/92) which can find orphans.

```
>>> intf.cable_terminations.all()
<RestrictedQuerySet [<CableTermination: Cable #51 to ether8>]>
```

If you want to trace where the cable goes then there's `intf.link_peers` but in general there could be multiple endpoints to follow (since one cable can have multiple terminations at the other end). `intf.trace()` will give you a detailed trace of the path which you can walk to find all the cables en route:

```
>>> intf.connected_endpoints
[<Interface: eno1>]
>>> intf.link_peers
[<FrontPort: F3>]
>>> intf.trace()
[([<Interface: ether8>], [<Cable: #51>], [<FrontPort: F3>]), ([<RearPort: R3>], [<Cable: #54>], [<Interface: eno1>])]
```

Again beware that you can't blindly delete call cables in the path because they *might* be connecting other things (e.g. one cable may link device 1 interface A and device 2 interface B at one end, to device 3 interface C and device 4 interface D at the other end). Such cables are uncommon, but typically used for things like QSFP+ breakout cables.

**Reply by domainzero** [reply]
Very much appreciate the in-depth and well thought out answer instead of just throwing my question into ChatGPT. :smile: 
I probably don't need to actually recurse all the way down the line. If I need to remove cables from a patch panel I can do it manually.

This looks like enough for me to get started.

Thanks a lot!
